test/DoorLock.purs

```diff
@@ -0,0 +1,87 @@
+module Test.DoorLock
+       ( startLink
+       , State
+       )
+       where
+
+
+import Prelude
+
+import Data.Either (Either(..))
+import Data.Maybe (Maybe(..))
+import Effect (Effect)
+import Erl.Atom (atom)
+import Erl.Data.List (nil, (:))
+import Pinto.GenStatem (StateEnterAction(..), StateEnterResult(..), StatemType, Timeout(..), TimeoutAction(..), Running(..))
+import Pinto.GenStatem as Statem
+import Pinto.Types (InstanceRef(..), RegistryName(..), ServerPid, StartLinkResult, crashIfNotStarted)
+import Unsafe.Coerce (unsafeCoerce)
+
+-- type Cont = Void
+-- type Stop = Void
+-- data Msg = SetValue Int
+
+data EventContent
+  = DoorOpenTooLong
+
+data State
+  = Locked
+  | UnlockedClosed
+  | UnlockedOpen
+
+type Data = {}
+
+type DoorLockType = StatemType State Data
+
+data AuditEvent
+  = AuditDoorUnlocked
+  | AuditDoorOpened
+  | AuditDoorClosed
+  | AuditDoorLocked
+
+
+name :: RegistryName DoorLockType
+name = Local $ atom "doorLock"
+
+
+startLink :: Effect (ServerPid DoorLockType)
+startLink = do
+  crashIfNotStarted <$> statem
+  where
+    spec :: Statem.Spec State Data
+    spec = Statem.mkSpec
+
+    statem :: Effect (StartLinkResult DoorLockType)
+    statem = Statem.startLink spec
+    -- init =
+    --   let
+    --     initialState = Locked
+    --     initialData = { }
+    --   in
+    --     pure $ Right $ InitOk initialState initialData
+
+    handleEvent = unsafeCoerce unit
+
+    handleEnter Locked UnlockedClosed currentData = do
+      audit AuditDoorUnlocked
+      pure $ Right $ KeepStateAndData
+
+    handleEnter UnlockedOpen UnlockedClosed currentData = do
+      audit AuditDoorClosed
+      pure $ Right $ KeepStateAndData
```

Comment:
Again, this whole Right/Left thing around our ADTs doesn't look right

Reply:
This is an older commit, the Right/Left distinction here was just copy/pasta, it's not there in the current version, except in the init function.

As to whether it should be there in the init function and in the StartLinkResult, I ended up bowing out of that debate, others had fairly strong opinions there and I was pretty much on the fence about the whole thing.

Reply:
:ok:

Reply:
er, :ok_hand:  even